### DISCUSSION TITLE
Details about the Function.dat.data

### DISCUSSION THREAD

**Discussion by tu1620808124** [firedrakeproject/firedrake#2890] [Firedrake support] [answered]
Hello, 
I want to know the detail of Function.dat.data,
```
from firedrake import *
import numpy as np
import math,sys
nx = 2
mesh = UnitSquareMesh(nx,nx,quadrilateral=True)
x = SpatialCoordinate(mesh)
space = FunctionSpace(mesh, "DQ", 1)
v = TestFunction(space)
f = Function(space).interpolate(sin(x[0]))
print(f.dat.data)
print(f.at(0., 0.))
```
the results
```
[0.10546593 0.10546593 0.38419668 0.38419668 0.5693068  0.5693068
 0.77979455 0.77979455 0.10546593 0.10546593 0.38419668 0.38419668
 0.5693068  0.5693068  0.77979455 0.77979455]
0.003443394450575697
```
Function.dat.data is not an array at physical points seems the gauss points. but when the ```A = assemble(f*v*dx)``` , what is the A.dat.data meaning?

**Comment by stephankramer** [comment, marked as answer]
`A.dat.data` are the coefficients w.r.t. the basis which for `DQ` indeed corresponds to values in nodes at physical locations, it's just not the physical locations you are expecting. The default choice of nodes in Firedrake for discontinuous elements correspond to the Gauss Legendre points, which does not include the vertices. If you want the more "standard" choice of nodes that includes the vertices, with equispaced notes inbetween for higher order, use:
```
ele = FiniteElement("DQ", mesh.ufl_cell(), 1, variant='equispaced')
space = FunctionSpace(mesh, ele)
```
